R/vad.R

```diff
@@ -0,0 +1,183 @@
+#'      the `ggplot2` package. Labels could, for example, be modified using [ggplot2::labs()] or [ggplot2::ggtitle()].
+#'      Using [ggplot2::theme()] the visual appearance can easily be modified. To do this the regular [+][ggplot2::+.gg]
+#'      syntax can be used. In the examples this is demonstrated using scale.
+#'
+#' @export
+#' @examples
+#' pvolfile <- system.file("extdata", "volume.h5", package = "bioRad")
+#' example_pvol <- read_pvolfile(pvolfile)
+#' # VAD plots can be created for polar volumes alone
+#' vad(example_pvol,
+#'   range = c(5000, 30000), height = c(200, 400)
+#' )
+#' # It is also possible to plot one height or more height bins from a `vp`.
+#' # Many visual aspects can be controlled through the function arguments
+#' # or through adding `ggplot`
+#' vp <- calculate_vp(pvolfile)
+#' vad(example_pvol,
+#'   vp = vp,
+#'   height = 400,
+#'   plotting_geom_args=list(ggplot2::aes(color=ZDR))
+#' ) + ggplot2::scale_color_gradient2()
+#' vad(example_pvol,
+#'   vp = vp, height=c(400,1200),
+#'   plotting_geom=ggplot2::geom_bin2d,
+#'   annotate="sdvvp: {round(sd_vvp,2)} [m/s]",
+#' ) + ggplot2::scale_fill_viridis_c()
+vad <- function(x, ...) {
+  UseMethod("vad", x)
+}
+#' @rdname vad
+#' @export
+vad.pvol <- function(x, vp = NULL,..., range = NULL, height = NULL,
+                     range_gate_filter = DBZH < 20 & RHOHV < .95,
+                     plotting_geom=ggplot2::geom_point,
+                     plotting_geom_args=list(),
+                     annotate="{round(ff,1)} m/s, {round(dd)}\u00B0",
+                     annotation_size=4,
+                     vp_color="red"){
+  assertthat::assert_that(is.pvol(x))
+  if (!is.null(vp)) {
+    vp_df <- as.data.frame(vp) |> dplyr::mutate(
+      height_bin = glue::glue("{height}-{height + vp$attributes$where$interval} [m]"),
+      height_bin = factor(.data$height_bin, levels = .data$height_bin)
+    )
+
+    assertthat::assert_that(
+      is.null(range),
+      msg = "When specifying a vp the range is taken from there. Thus no range should be provided"
+    )
+    range <- c(vp$attributes$how$minrange, vp$attributes$how$maxrange) * 1000
+    if (length(height) == 1) {
+      height <- max(vp$data$height[vp$data$height <= height]) + c(0, vp$attributes$where$interval)
+    }
+
+    if (!is.null(height)) {
+      # if a height profile has been provided we only plot those curves from a vp
+      vp_df <- vp_df[(vp_df$height + vp$attributes$where$interval) > min(height) & vp_df$height < max(height), ]
+    }
+  }
+  assertthat::assert_that(
+    is.null(range) ||
+      (is.numeric(range) && length(2))
+  )
+  assertthat::assert_that(
+    is.null(height) ||
+      (is.numeric(height) && length(2)) ||
+      (!is.null(vp) && is.numeric(height))
+  )
+  if (is.null(range)) {
+    range <- c(-Inf, Inf)
+  }
+  if (is.null(height)) {
+    height <- c(-Inf, Inf)
+  }
+  data <-
+    mapply(SIMPLIFY = F,
+      cbind,
+      lapply( lapply(x$scans, scan_to_spatial),
```

Comment:
I think you need an option to select an elevation scan range. For example, for a 90 degree bird bath scan this won't work. See also my comment below on the effect of elevation angle on detected radial velocity